devtools/conda-envs/psi4.yaml

```diff
@@ -8,7 +8,7 @@ channels:
 dependencies:
 
   # Base depends
-  - python>=3.8
+  - python =3.9
```

Comment:
I don't think we need to pin the python version, this should be handled in the CI.yaml where we define the versions of python we want to test on?

Reply:
Hmm, that might be true, at least for the CI. When I had >=3.8 it failed locally because it installed 3.10 and we ran into the issue with `collections` I mentioned in the basic commit. Mine is failing locally anyway because of PCMSolver, so we can try it without the pin if you want.

Reply:
I guess we could do something like >=3.8, <3.10, but I think psi4 1.8 also requires >=3.9, so the end result should be 3.9 one way or another.

Reply:
I'm fine with unpinning for CI, but should we add additional conda-envs for local testing then? If we unpin the Python version, anyone trying to develop locally will get a broken environment.

Also if we unpin here, my earlier change for the basic tests needs to be unpinned as well, which is, again, fine with me. I just want to be able to set up a dev environment locally when working on this, and I think it would be best if those steps were encoded in the repo itself. Maybe that point is irrelevant once we can actually handle Python 3.10, though.

Reply:
What we've sorta settled on is having multiple environment YAML files for different environments that need significantly different sets of upstreams, i.e.

* https://github.com/openforcefield/openff-models/tree/main/devtools/conda-envs
* https://github.com/openforcefield/openff-forcefields/tree/main/devtools/conda-envs

and having the [workflow YAML](https://github.com/openforcefield/openff-forcefields/blob/e3b4387c3898412cdb5c68533ac81981bf7dd316/.github/workflows/CI.yml#L51-L52) be the single source of truth for Python versions in addition to constraints on upstream dependencies.

What's worked well for us is to assume individual developers' environments will diverge from what CI uses - I bring in a lot of tools that CI doesn't need, for example - but that everybody should start from there. After all, getting CI passing is a (soft) requirement for changes to be accepted. Sometimes I track a development environment for my own purposes, but this is neither something that CI sees nor something that I'd require other developers to start from.

I was able to solve the environment from this branch pretty quickly (`CONDA_SUBDIR=osx-64 mamba env create --file devtools/conda-envs/basic.yaml`) and get Python 3.9. I might be missing something, but is there an unmet need that would be satisfied by tracking a separate testing environment meant for local installs?

Reply:
> I was able to solve the environment from this branch pretty quickly (CONDA_SUBDIR=osx-64 mamba env create --file devtools/conda-envs/basic.yaml) and get Python 3.9

Yes, the basic environment is currently pinned on 3.9 too, so I would expect that you did get Python 3.9. If I take that basic.yaml and replace `python =3.9` with `python >=3.8` as proposed here and run `mamba env create -f devtools/conda-envs/basic.yaml`, I end up with Python 3.10.12, which means the tests will fail with the import error from collections.

> I might be missing something, but is there an unmet need that would be satisfied by tracking a separate testing environment meant for local installs?

Personally, I think it makes more sense *not* to have a separate local environment, but to ensure the CI environment is usable for local testing, which is the case with =3.9 and not the case without it. How else are you supposed to know if your changes pass the tests on CI without pushing them?

The separate environment suggestion was just an idea since it didn't seem like we wanted to make the CI environment usable locally.

I'm new to this and probably misunderstanding something. I just want a way to know that my local changes are going to pass CI before pushing them. I assumed that this meant I needed to be running in the same environment as the CI, but I am open to other options since y'all have obviously been managing this since before I joined.

More broadly, it feels to me that we should have something like fully-pinned environments that we know work. Nearly everything I've tried to do so far has required lots of environment debugging even to get code to run, much less to get tests to pass. Maybe these shouldn't be the normal CI environments, but I just feel like they should exist and be tracked somewhere with the code they run.

Reply:
I hadn't thought about this before. Given how useful it is to be able to use `mamba env create -f devtools/conda-envs/xxx.yaml` to make developer builds, I think it could be useful to BOTH:

* continue pinning exact versions of python in the CI build yamls (in the `.github` directory), AND
* specify a python version range in the yamls in `conda_envs` directory, for local installs

In this case, the python version range in the `conda_envs` yamls would just be `=3.9` (or if we wanted to encode intent, `>=3.9, <3.10`). We can re-evaluate this at each release since the CI yamls are usually where we look to determine packaging requirements, so we'll be motivated to resolve the `Mapping` thing every time we have to downpin to `python<3.10` in the package requirements.

I don't think we should have a separate yaml for local development builds - it seems really likely that it'd fall out of sync and we'd stop using it.

Reply:
After some thought, Matt's convinced me that it's best to only specify python version pins in the workflow file, and not in the conda-envs yaml. While it would solve the "building a dev environment gives me a broken env with python 3.10" problem _now_, it will create a problem in just a few hours when we're trying to fix python 3.10 compatibility and the pin in our conda-envs yaml is actively PREVENTING us from making a dev environment with python>=3.10. 

Neither solution is perfect, but incompatibility with the newer versions of python should be a rare occurrence, and it'll be USEFUL if dev builds warn us about incompatibilities with new python versions, so we can get them onto the issue tracker ASAP.

Reply:
In that case, should Python 3.10 also be specified in the workflow YAML? I see in the workflow Matt linked from the openff-forcefields repo that the Python version matrix contains both Python 3.9 and 3.10, but in this repo it only has 3.9. That would probably be a better way to catch incompatibilities than someone installing a new dev environment. And Python 3.11 was release 10 months ago, are there any plans for supporting/checking that too?

Reply:
The goal should be to support and test all supported versions within NEP 29, which currently includes 3.9-3.11. I have no idea if this will work on each of those versions right now

https://numpy.org/neps/nep-0029-deprecation_policy.html#support-table

Reply:
Sounds like we should add 3.10 and 3.11 to the workflow then.

Reply:
> In that case, should Python 3.10 also be specified in the workflow YAML?

Yeah, with more dev time + CI resources we'd have all of our packages tested on 3.9, 3.10, and 3.11.